clients/registry-amqp/src/main/java/org/eclipse/hono/client/registry/amqp/ProtonBasedTenantClient.java

```diff
@@ -210,10 +220,59 @@ protected String getKey(final String tenantId) {
                         // otherwise execute the request
                         () -> serviceRequest.get()
                                 // make sure to put the response to the cache (if applicable)
-                                .onSuccess(tenantResult -> addToCache(responseCacheKey, tenantResult))
+                                .onSuccess(tenantResult -> addResultToCache(responseCacheKey, tenantResult))
                                 // and again remove the result promise so that a subsequent request will be executed again
                                 .onComplete(ar -> pendingRequests.remove(responseCacheKey))
                                 .onComplete(resultPromise));
         return resultPromise.future();
     }
+
+    private void addResultToCache(final AnnotatedCacheKey<?> responseCacheKey,
+            final TenantResult<TenantObject> tenantResult) {
+
+        if (isCachingEnabled()) {
+            Optional.ofNullable(tenantResult.getPayload())
+                    .ifPresent(p -> responseCacheKey.putAttribute(ATTRIBUTE_KEY_TENANT_ID, p.getTenantId()));
+
+            addToCache(responseCacheKey, tenantResult);
+        }
+    }
+
+    private void removeResultFromCache(final String tenantId) {
+        removeFromCacheByPattern(key -> ((AnnotatedCacheKey<?>) key)
```

Comment:
what about the cache entries where the key is the tenant ID?

Reply:
The cache entries where the key is the tenant ID are also removed by this because every cache key contains this metadata. I added comments to make it clearer. We need to search by the pattern anyways to find the cache entries with the principal as the key.